src/structs/multiplexer.rs

```diff
@@ -0,0 +1,376 @@
+            }
+
+            copy_start += amount_to_copy;
+        }
+
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        if self.buffer.len() > 0 {
+            let mut new_buffer = Vec::with_capacity(WRITE_BUFFER_SIZE);
+            swap(&mut new_buffer, &mut self.buffer);
+
+            self.sender
+                .send(Message::WriteBlock(self.thread_id, new_buffer))
+                .unwrap();
+        }
+        Ok(())
+    }
+}
+
+/// Given an arbitrary writer, this function will launch the given number of threads and call the processor function
+/// on each of them, and collect the output written by each thread to the writer in blocks identified by the thread_id.
+///
+/// This output stream can be processed by multiple_read to get the data back, using the same number of threads.
+pub fn multiplex_write<WRITE, FN, RESULT>(
+    writer: &mut WRITE,
+    num_threads: usize,
+    processor: FN,
+) -> Result<Vec<RESULT>>
+where
+    WRITE: Write,
+    FN: Fn(&mut MultiplexWriter, usize) -> Result<RESULT> + Send + Copy,
+    RESULT: Send,
+{
+    let mut thread_results = Vec::<Option<Result<RESULT>>>::new();
+
+    for _i in 0..num_threads {
+        thread_results.push(None);
+    }
+
+    rayon::in_place_scope(|s| -> Result<()> {
+        let (tx, rx) = channel();
+
+        for (thread_id, result) in thread_results.iter_mut().enumerate() {
+            let cloned_sender = tx.clone();
+
+            let mut thread_writer = MultiplexWriter {
+                thread_id: thread_id as u8,
+                sender: cloned_sender,
+                buffer: Vec::with_capacity(WRITE_BUFFER_SIZE),
+            };
+
+            let mut f = move || -> Result<RESULT> {
+                let r = processor(&mut thread_writer, thread_id)?;
+
+                thread_writer.flush().context(here!())?;
+
+                thread_writer.sender.send(Message::Eof).context(here!())?;
+                Ok(r)
+            };
+
+            s.spawn(move |_| {
+                *result = Some(f());
+            });
+        }
+
+        // drop the sender so that the channel breaks when all the threads exit
+        drop(tx);
+
+        // wait to collect work and done messages from all the threads
+        let mut threads_left = num_threads;
+
+        while threads_left > 0 {
+            let value = rx.recv().context(here!());
+            match value {
+                Ok(Message::Eof) => {
+                    threads_left -= 1;
+                }
+                Ok(Message::WriteBlock(thread_id, b)) => {
+                    let l = b.len() - 1;
+
+                    writer.write_u8(thread_id).context(here!())?;
+                    writer.write_u8((l & 0xff) as u8).context(here!())?;
+                    writer.write_u8(((l >> 8) & 0xff) as u8).context(here!())?;
+                    writer.write_all(&b[..]).context(here!())?;
+                }
+                Err(_) => {
+                    break;
```

Comment:
Could you please explain? I see that on error we are exiting the loop but do not return the error? Are we relying on one of the threads to return error in this case?

Also, is rayon::in_place_scope execution completes only after all the spawned threads complete, that's why we do not have any kind of "join" here?